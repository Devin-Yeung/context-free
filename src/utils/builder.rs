use crate::utils::symbols;
use bnf::{Grammar, Term};
use once_cell::sync::OnceCell;
use std::collections::{HashMap, HashSet};

pub fn epsilon() -> &'static Term {
    static EPSILON: OnceCell<Term> = OnceCell::new();
    EPSILON.get_or_init(|| Term::Terminal(String::from("ε")))
}

fn dollar() -> &'static Term {
    static DOLLAR: OnceCell<Term> = OnceCell::new();
    DOLLAR.get_or_init(|| Term::Terminal(String::from("$")))
}

pub struct FirstFollowBuilder<'grammar> {
    pub(crate) inner: HashMap<&'grammar Term, HashSet<&'grammar Term>>,
}

impl<'grammar> FirstFollowBuilder<'grammar> {
    pub(crate) fn new(grammar: &'grammar Grammar) -> FirstFollowBuilder<'grammar> {
        let mut inner = HashMap::new();

        // initialize the table
        symbols(&grammar)
            .into_iter()
            .filter(|term| term != &epsilon()) // epsilon is a special non-terminal
            .for_each(|term| {
                inner.insert(term, HashSet::new());
            });

        FirstFollowBuilder { inner }
    }

    // Insert term to First(x)
    ///
    /// return true if the First(x) changes
    /// otherwise return false
    pub(crate) fn insert_term(&mut self, x: &'grammar Term, term: &'grammar Term) -> bool {
        // First(x)
        let first_x = self.inner.get_mut(x).unwrap();

        // Insert term to First(x)
        let before = first_x.len();
        first_x.insert(term);
        let after = first_x.len();

        // check if set changes
        before != after
    }

    // Insert epsilon to First(x)
    pub(crate) fn insert_epsilon(&mut self, x: &'grammar Term) -> bool {
        self.insert_term(x, epsilon())
    }

    // Insert dollar to Follow(x)
    pub(crate) fn insert_dollar(&mut self, x: &'grammar Term) -> bool {
        self.insert_term(x, dollar())
    }

    /// First(x)
    pub(crate) fn first(&self, x: &Term) -> HashSet<&'grammar Term> {
        self.inner
            .get(x)
            .map_or_else(|| HashSet::new(), |set| set.clone())
    }

    /// Follow(x)
    pub(crate) fn follow(&self, x: &Term) -> HashSet<&'grammar Term> {
        self.inner
            .get(x)
            .map_or_else(|| HashSet::new(), |set| set.clone())
    }

    pub(crate) fn insert_set(&mut self, x: &'grammar Term, set: HashSet<&'grammar Term>) -> bool {
        // First(x)
        let first_x = self.inner.get_mut(x).unwrap();

        // Insert set into First(x)
        let before = first_x.len();
        first_x.extend(set);
        let after = first_x.len();

        // check if set changes
        return before != after;
    }

    /// Insert First(y) \ { ε } into First(x)
    ///
    /// return true if the First(x) changes
    /// otherwise return false
    pub(crate) fn insert_first_no_epsilon(&mut self, x: &'grammar Term, y: &'grammar Term) -> bool {
        // First(y)
        let mut first_y = self.first(y);
        // First(y) \ { ε }
        first_y.remove(epsilon());
        // Insert First(y) \ { ε } into First(x)
        self.insert_set(x, first_y)
    }

    /// Insert Follow(y) into Follow(x)
    pub(crate) fn insert_follow(&mut self, x: &'grammar Term, y: &'grammar Term) -> bool {
        // Follow(y)
        let follow_y = self.follow(y);
        println!("Insert {:?} to Follow({})", follow_y, x);
        // Insert Follow(y) into Follow(x)
        self.insert_set(x, follow_y)
    }

    pub(crate) fn build(self) -> HashMap<&'grammar Term, HashSet<&'grammar Term>> {
        self.inner
    }
}
