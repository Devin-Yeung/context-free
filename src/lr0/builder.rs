use crate::lr0::core::{LR0Closure, LR0Item, LR0ItemSet};
use crate::utils::symbols;
use bnf::{Grammar, Production, Term};
use std::collections::{HashMap, VecDeque};

pub struct LR0Builder<'grammar> {
    grammar: &'grammar Grammar,
    closures: Vec<LR0ItemSet<'grammar>>,
    transitions: HashMap<(usize, &'grammar Term), usize>,
}

impl<'grammar> LR0Builder<'grammar> {
    pub fn new(grammar: &'grammar Grammar) -> LR0Builder<'grammar> {
        LR0Builder {
            grammar,
            closures: Vec::new(),
            transitions: HashMap::new(),
        }
    }

    pub fn build(mut self, augmentation: &'grammar Production) -> LR0Closure {
        let initial = LR0ItemSet::from_iter(vec![LR0Item::from_production(augmentation).unwrap()]);
        self.build_closure(&initial);
        self.build_transition();
        LR0Closure {
            closures: self.closures,
            transitions: self.transitions,
        }
    }

    fn build_closure(&mut self, initial: &LR0ItemSet<'grammar>) {
        let mut waiting = VecDeque::from([initial.closure(self.grammar)]);
        self.closures.push(initial.closure(self.grammar));

        while !waiting.is_empty() {
            let cur = waiting.pop_front().unwrap();
            symbols(self.grammar).iter().for_each(|term| {
                let goto = cur.goto(self.grammar, term);
                if !goto.items.is_empty() && !self.contains(&goto) {
                    self.closures.push(goto);
                    waiting.push_back(self.closures.last().unwrap().clone());
                }
            });
        }
    }

    fn build_transition(&mut self) {
        for closure in &self.closures {
            for term in symbols(self.grammar) {
                let goto = closure.goto(self.grammar, term);
                if goto.items.is_empty() {
                    continue;
                }
                let goto_index = self.index_of(&goto);
                let cur_index = self.index_of(closure);
                self.transitions.insert((cur_index, term), goto_index);
                println!("goto(I_{}, {}) = I_{}", cur_index, term, goto_index);
            }
        }
    }

    fn contains(&self, set: &LR0ItemSet<'grammar>) -> bool {
        self.closures.iter().any(|v| v == set)
    }

    fn index_of(&self, set: &LR0ItemSet<'grammar>) -> usize {
        self.closures.iter().position(|v| v == set).unwrap()
    }
}

#[cfg(test)]
mod tests {
    use crate::lr0::core::LR0Closure;
    use bnf::Production;
    use std::str::FromStr;

    #[test]
    fn it_works() {
        let grammar = r#"
        <E'> ::= <E>
        <E> ::= <E> '+' <T> | <T>
        <T> ::= <T> '*' <F> | <F>
        <F> ::= '(' <E> ')' | 'id'
        "#
        .parse()
        .unwrap();
        let augmentation = Production::from_str("<E'> ::= <E>").unwrap();

        assert_eq!(
            LR0Closure::new(&grammar, &augmentation).closures().len(),
            12
        );
    }

    #[test]
    fn object_clause() {
        let grammar = r#"
        <S> ::= <O> 'v' <C>
        <O> ::= 'n'
        <C> ::= <S>
        <C> ::= 'n'
        "#
        .parse()
        .unwrap();
        let augmentation = Production::from_str("<S'> ::= <S>").unwrap();

        assert_eq!(LR0Closure::new(&grammar, &augmentation).closures().len(), 8);
    }
}
