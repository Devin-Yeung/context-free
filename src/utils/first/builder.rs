use crate::utils::first::First;
use crate::utils::{epsilon, symbols};
use bnf::{Grammar, Production, Term};
use std::cell::RefCell;
use std::collections::{HashMap, HashSet};

pub struct FirstBuilder<'grammar> {
    pub(crate) grammar: &'grammar Grammar,
    pub(crate) first: RefCell<HashMap<&'grammar Term, HashSet<&'grammar Term>>>,
    pub(crate) lookup: HashMap<&'grammar Term, &'grammar Production>,
}

impl<'grammar> FirstBuilder<'grammar> {
    pub(crate) fn new(grammar: &'grammar Grammar) -> FirstBuilder<'grammar> {
        let mut first = HashMap::new();

        let lookup = grammar
            .productions_iter()
            .map(|production| (&production.lhs, production))
            .collect::<HashMap<_, _>>();

        // initialize the table
        symbols(grammar)
            .filter(|term| term != &epsilon()) // epsilon is a special non-terminal
            .for_each(|term| {
                first.insert(term, HashSet::new());
            });

        let first = RefCell::new(first);

        FirstBuilder {
            grammar,
            first,
            lookup,
        }
    }

    fn build_first(&mut self) {
        symbols(self.grammar)
            .filter(|term| term != &epsilon())
            .for_each(|t| {
                match t {
                    Term::Terminal(s) => {
                        // Rule1: If X is a terminal, then First(X) = { X }
                        self.insert_term(t, t);
                        println!("Rule1: Push {} to First({})", s, t);
                    }
                    Term::Nonterminal(_) => { /* skip */ }
                };

                if self.produce_epsilon(t) {
                    // Rule2: If X is an ε-production, then add ε to First(X)
                    self.insert_epsilon(t);
                    println!("Rule2: Push ε to First({})", t);
                }
            });

        loop {
            let mut changed = false;

            symbols(self.grammar)
                .filter(|term| matches!(*term, Term::Nonterminal(_)))
                .for_each(|lhs| {
                    println!("===> Checking Symbol: {}", lhs);
                    let production = self.lookup.get(lhs).unwrap();
                    // Rule3: If X is a non-terminal and X → Y1 Y2 ... Yk,
                    // then add First(Y1) ∖ {ε} to First(X)
                    for expr in production.rhs_iter() {
                        for term in expr
                            .terms_iter()
                            .filter(|term| term != &&Term::Terminal("ε".to_string()))
                        {
                            // First(Y1) ∖ {ε} to First(X)
                            changed |= self.insert_first_no_epsilon(&production.lhs, term);
                            println!(
                                "Rule3/4: Push First({}) \\ ε to First({})",
                                term, production.lhs
                            );
                            // terminate (check next expression) if X does NOT produce ε
                            if !self.produce_epsilon(term) {
                                println!("{} does NOT produce ε", term);
                                break;
                            }
                        }
                        // Rule 5: If X is a non-terminal and X -> Y1 Y2 ... Yk,
                        // and First(Yi) produce ε for all i, then add ε to First(X)
                        if expr.terms_iter().all(|term| self.produce_epsilon(term)) {
                            println!("Rule5: Push ε to First({})", production.lhs);
                            changed = self.insert_epsilon(&production.lhs);
                        }
                    }
                });

            if !changed {
                println!("Unchanged, break!");
                break;
            }
        } // End of loop
    }

    fn produce_epsilon(&self, term: &Term) -> bool {
        let production = self.lookup.get(&term);
        if production.is_none() {
            return false;
        }

        let production = production.unwrap();

        match &production.lhs {
            Term::Terminal(t) => {
                if t == "ε" {
                    return true;
                }
            }
            Term::Nonterminal(nt) => {
                if nt == "ε" {
                    return true;
                }
            }
        }

        production
            .rhs_iter()
            .map(|expr| {
                expr.terms_iter().all(|term| match term {
                    Term::Terminal(t) => t == "ε",
                    Term::Nonterminal(nt) => nt == "ε",
                })
            })
            .any(|v| v)
    }

    // Insert term to First(x)
    ///
    /// return true if the First(x) changes
    /// otherwise return false
    fn insert_term(&self, x: &'grammar Term, term: &'grammar Term) -> bool {
        let mut first = self.first.borrow_mut();
        // First(x)
        dbg!(x);
        let first_x = first.get_mut(x).unwrap();

        // Insert term to First(x)
        let before = first_x.len();
        first_x.insert(term);
        let after = first_x.len();

        // check if set changes
        before != after
    }

    // Insert epsilon to First(x)
    fn insert_epsilon(&self, x: &'grammar Term) -> bool {
        self.insert_term(x, epsilon())
    }

    /// First(x)
    fn first(&self, x: &Term) -> HashSet<&'grammar Term> {
        self.first
            .borrow()
            .get(x)
            .map_or_else(HashSet::new, |set| set.clone())
    }

    fn insert_set(&self, x: &'grammar Term, set: HashSet<&'grammar Term>) -> bool {
        let mut first = self.first.borrow_mut();
        // First(x)
        let first_x = first.get_mut(x).unwrap();

        // Insert set into First(x)
        let before = first_x.len();
        first_x.extend(set);
        let after = first_x.len();

        // check if set changes
        before != after
    }

    /// Insert First(y) \ { ε } into First(x)
    ///
    /// return true if the First(x) changes
    /// otherwise return false
    fn insert_first_no_epsilon(&self, x: &'grammar Term, y: &'grammar Term) -> bool {
        // First(y)
        let mut first_y = self.first(y);
        // First(y) \ { ε }
        first_y.remove(epsilon());
        // Insert First(y) \ { ε } into First(x)
        self.insert_set(x, first_y)
    }

    pub(crate) fn build(mut self) -> First<'grammar> {
        self.build_first();
        First {
            first: self.first.into_inner(),
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::utils::first::builder::FirstBuilder;
    use bnf::Term;

    #[test]
    fn first() {
        let grammar = r#"
        <E> ::= <T> <E'>
        <E'> ::= '+' <T> <E'> | 'ε'
        <T> ::= <F> <T'>
        <T'> ::= '*' <F> <T'> | 'ε'
        <F> ::= '(' <E> ')' | 'id'
        "#
        .parse()
        .unwrap();

        let first = FirstBuilder::new(&grammar).build();
        first.first.iter().for_each(|(lhs, rhs)| match lhs {
            Term::Terminal(_) => {
                assert_eq!(rhs.len(), 1)
            }
            Term::Nonterminal(_) => {
                assert_eq!(rhs.len(), 2)
            }
        })
    }
}
