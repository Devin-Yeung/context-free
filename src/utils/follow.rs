use crate::utils::builder::{epsilon, FirstFollowBuilder};
use crate::utils::first::First;
use bnf::{Grammar, Production, Term};
use std::collections::{HashMap, HashSet};

pub struct Follow<'grammar> {
    grammar: &'grammar Grammar,
    lookup: HashMap<&'grammar Term, &'grammar Production>,
    first: HashMap<&'grammar Term, HashSet<&'grammar Term>>,
}

impl<'grammar> Follow<'grammar> {
    pub fn new(grammar: &'grammar Grammar) -> Follow<'grammar> {
        let lookup = grammar
            .productions_iter()
            .map(|production| (&production.lhs, production))
            .collect::<HashMap<_, _>>();

        let first = First::new(grammar).first();

        Follow {
            grammar,
            lookup,
            first,
        }
    }

    pub fn first_produce_epsilon(&self, term: &Term) -> bool {
        self.first
            .get(term)
            .map_or(false, |first| first.contains(&epsilon()))
    }

    pub fn follow(
        &self,
        start: &'grammar Term,
    ) -> HashMap<&'grammar Term, HashSet<&'grammar Term>> {
        let mut follow = FirstFollowBuilder::new(&self.grammar);
        // Rule 1: If X is a start symbol, then Follow(X) = { $ }
        follow.insert_dollar(start);
        println!("Rule 1: Push $ to Follow({})", start.to_string());

        loop {
            let mut changed = false;
            for production in self.grammar.productions_iter() {
                println!("==> Checking production {}", production.lhs.to_string());
                for expr in production.rhs_iter() {
                    // Rule 2 checking
                    let mut prev: Option<&'grammar Term> = None;
                    for (idx, term) in expr
                        .terms_iter()
                        .collect::<Vec<_>>()
                        .iter()
                        .rev()
                        .enumerate()
                    {
                        if idx != 0 && matches!(term, Term::Nonterminal(_)) {
                            let mut first_yi = self
                                .first
                                .get(prev.unwrap())
                                .map_or_else(|| HashSet::new(), |set| set.clone());
                            first_yi.remove(epsilon());
                            println!(
                                "Rule 2: Push First({}) \\ ε to Follow({})",
                                prev.unwrap().to_string(),
                                term.to_string()
                            );
                            changed |= follow.insert_set(term, first_yi);
                        }

                        prev = Some(term);
                    } // Rule 2 checking End

                    // Rule 3 checking
                    for term in expr.terms_iter().collect::<Vec<_>>().iter().rev() {
                        // if
                        if matches!(term, Term::Terminal(_)) {
                            break;
                        }
                        // Rule 3: If X -> Y1 Y2 ... Yk,
                        // then add Follow(X) to Follow(Yk)
                        println!(
                            "Rule 3: Push Follow({}) to Follow({})",
                            production.lhs.to_string(),
                            term.to_string()
                        );
                        changed |= follow.insert_follow(term, &production.lhs);

                        if !self.first_produce_epsilon(term) {
                            println!("{} does not produce ε, break", term.to_string());
                            break;
                        }
                    } // Rule 3 checking End
                }
            }

            if !changed {
                break;
            }
        }

        follow.build()
    }
}

#[cfg(test)]
mod tests {
    use crate::utils::follow::Follow;
    use bnf::{Grammar, Term};
    use std::collections::{HashMap, HashSet};
    use std::str::FromStr;

    pub fn grammar() -> Grammar {
        let input = r#"
        <E> ::= <T> <E'>
        <E'> ::= '+' <T> <E'> | 'ε'
        <T> ::= <F> <T'>
        <T'> ::= '*' <F> <T'> | 'ε'
        <F> ::= '(' <E> ')' | 'id'
        "#;
        let grammar: Grammar = input.parse().unwrap();
        grammar
    }

    fn get_follow<'a>(
        follow: &HashMap<&'a Term, HashSet<&'a Term>>,
        term: &str,
    ) -> HashSet<&'a Term> {
        let term = Term::from_str(term).unwrap();
        follow.get(&term).unwrap().clone()
    }

    #[test]
    fn it_works() {
        let grammar = grammar();
        let start = Term::Nonterminal("E".to_string());
        let follow = Follow::new(&grammar).follow(&start);

        assert_eq!(get_follow(&follow, "<E>").len(), 2);
        assert_eq!(get_follow(&follow, "<E'>").len(), 2);
        assert_eq!(get_follow(&follow, "<T>").len(), 3);
        assert_eq!(get_follow(&follow, "<T'>").len(), 3);
        assert_eq!(get_follow(&follow, "<F>").len(), 4);
    }
}
