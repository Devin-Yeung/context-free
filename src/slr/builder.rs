use crate::lr0::core::{LR0Closure, LR0Item};
use crate::slr::core::SLRInstruction;
use crate::slr::helper::IndexedGrammar;
use crate::utils::dollar;
use crate::utils::follow::Follow;
use bnf::{Grammar, Production, Term};
use std::cell::RefCell;
use std::collections::HashMap;
use std::fmt::{format, Display, Formatter};
use std::iter::{once, repeat};
use tabled::builder::Builder;
use tabled::Table;

pub struct SLRTable<'grammar> {
    grammar: IndexedGrammar<'grammar>,
    table: Vec<HashMap<&'grammar Term, SLRInstruction>>,
}

impl<'grammar> SLRTable<'grammar> {
    pub fn grammar_table(&self) -> Table {
        self.grammar.grammar_table()
    }

    pub fn parsing_table(&self) -> Table {
        let mut builder = Builder::default();

        let header = self
            .grammar
            .terminals()
            .chain(once(dollar()))
            .chain(self.grammar.non_terminals())
            .collect::<Vec<_>>();

        builder.set_header(header.iter().map(|t| t.to_string()));

        self.table.iter().enumerate().for_each(|(_, table)| {
            let row = header
                .iter()
                .map(|t| table.get(t).unwrap_or(&SLRInstruction::Empty))
                .collect::<Vec<_>>();
            builder.push_record(row);
        });

        builder.index().build()
    }
}

impl<'grammar> Display for SLRTable<'grammar> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.write_fmt(format_args!("Grammar: \n{}\n", self.grammar_table()))?;
        f.write_fmt(format_args!("Table: \n{}", self.parsing_table()))
    }
}

pub struct SLRTableBuilder<'grammar> {
    grammar: IndexedGrammar<'grammar>,
    follow: Follow<'grammar>,
    closure: LR0Closure<'grammar>,
    table: RefCell<Vec<HashMap<&'grammar Term, SLRInstruction>>>,
}

impl<'grammar> SLRTableBuilder<'grammar> {
    pub fn new(
        grammar: &'grammar Grammar,
        augmentation: &'grammar Production,
    ) -> SLRTableBuilder<'grammar> {
        let follow = Follow::new(grammar, &augmentation.lhs);
        let closure = LR0Closure::new(grammar, augmentation);
        let grammar = IndexedGrammar::new(grammar);
        let table = RefCell::new(
            repeat(HashMap::<&Term, SLRInstruction>::new())
                .take(closure.len())
                .collect(),
        );
        SLRTableBuilder {
            grammar,
            follow,
            closure,
            table,
        }
    }

    fn shift(&self, from: usize, via: &'grammar Term) {
        let to = self.closure.transition(from, via).unwrap(); // TODO: really not sure whether this unwrap is safe?
        println!("Shift: goto(I_{}, {}) = I_{}", from, via, to);
        let mut table = self.table.borrow_mut();
        table
            .get_mut(from)
            .unwrap()
            .insert(via, SLRInstruction::Shift(to));
    }

    fn reduce(&self, index: usize, lr0: &LR0Item) {
        let grammar_index = self.grammar.get_index_of(&lr0.rhs).unwrap();
        let prod = self.grammar.get(lr0.rhs).unwrap();
        let mut table = self.table.borrow_mut();
        for term in self.follow.follow_of(&prod.lhs).collect::<Vec<_>>() {
            table[index].insert(term, SLRInstruction::Reduce(grammar_index));
            println!(
                "Reduce: set (I_{}, {}) = r{}",
                index,
                term.to_string(),
                grammar_index
            );
        }
    }

    fn goto(&self, from: usize, via: &'grammar Term) {
        debug_assert!(matches!(via, Term::Nonterminal(_)));
        let to = self.closure.transition(from, via).unwrap(); // TODO: really not sure whether this unwrap is safe?
        println!("Goto: goto(I_{}, {}) = I_{}", from, via, to);
        let mut table = self.table.borrow_mut();
        table
            .get_mut(from)
            .unwrap()
            .insert(via, SLRInstruction::Goto(to));
    }

    pub fn build(self) -> SLRTable<'grammar> {
        self.closure.enumerate_lr0().for_each(|(i, lr0)| {
            // take a look at the char after the •
            match lr0.expect() {
                // lr0 expect non character, which is the form of A -> 𝛼 •
                None => {
                    /* Reduce */
                    self.reduce(i, lr0)
                }
                Some(t) => {
                    match t {
                        Term::Terminal(_) => self.shift(i, t),   /* Shift */
                        Term::Nonterminal(_) => self.goto(i, t), /* Goto */
                    }
                }
            }
        });
        SLRTable {
            grammar: self.grammar,
            table: self.table.into_inner(),
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::slr::builder::SLRTableBuilder;
    use bnf::Production;
    use std::str::FromStr;

    #[test]
    fn it_works() {
        let grammar = r#"
        <E'> ::= <E>
        <E> ::= <E> '+' <T> | <T>
        <T> ::= <T> '*' <F> | <F>
        <F> ::= '(' <E> ')' | 'id'
        "#
        .parse()
        .unwrap();

        let augmentation = Production::from_str("<E'> ::= <E>").unwrap();

        let builder = SLRTableBuilder::new(&grammar, &augmentation);
        let slr = builder.build();
        insta::assert_display_snapshot!(slr);
    }
}
