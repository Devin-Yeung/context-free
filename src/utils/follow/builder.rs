use crate::utils::first::First;
use crate::utils::follow::Follow;
use crate::utils::symbols;
use crate::utils::{dollar, epsilon};
use bnf::{Grammar, Term};
use log::debug;
use std::cell::RefCell;
use std::collections::{HashMap, HashSet};

pub struct FollowBuilder<'grammar> {
    pub(crate) grammar: &'grammar Grammar,
    pub(crate) follow: RefCell<HashMap<&'grammar Term, HashSet<&'grammar Term>>>,
    pub(crate) first: HashMap<&'grammar Term, HashSet<&'grammar Term>>,
}

impl<'grammar> FollowBuilder<'grammar> {
    pub(crate) fn new(grammar: &'grammar Grammar) -> FollowBuilder<'grammar> {
        let mut follow = HashMap::new();

        // initialize the table
        symbols(grammar)
            .filter(|term| term != &epsilon()) // epsilon is a special non-terminal
            .for_each(|term| {
                follow.insert(term, HashSet::new());
            });

        let follow = RefCell::new(follow);
        let first = First::new(grammar).first;

        FollowBuilder {
            grammar,
            follow,
            first,
        }
    }

    fn build_follow(&mut self, start: &'grammar Term) {
        // Rule 1: If X is a start symbol, then Follow(X) = { $ }
        self.insert_dollar(start);
        debug!("[Follow Builder] Rule 1: Push $ to Follow({})", start);

        loop {
            let mut changed = false;
            for production in self.grammar.productions_iter() {
                debug!("[Follow Builder] Checking production {}", production.lhs);
                for expr in production.rhs_iter() {
                    // Rule 2 checking
                    let mut prev: Option<&'grammar Term> = None;
                    for (idx, term) in expr
                        .terms_iter()
                        .collect::<Vec<_>>()
                        .iter()
                        .rev()
                        .enumerate()
                    {
                        if idx != 0 && matches!(term, Term::Nonterminal(_)) {
                            let mut first_yi = self
                                .first
                                .get(prev.unwrap())
                                .map_or_else(HashSet::new, |set| set.clone());
                            first_yi.remove(epsilon());
                            debug!(
                                "[Follow Builder] Rule 2: Push First({}) \\ ε to Follow({})",
                                prev.unwrap(),
                                term
                            );
                            changed |= self.insert_set(term, first_yi);
                        }

                        prev = Some(term);
                    } // Rule 2 checking End

                    // Rule 3 checking
                    for term in expr.terms_iter().collect::<Vec<_>>().iter().rev() {
                        // if
                        if matches!(term, Term::Terminal(_)) {
                            break;
                        }
                        // Rule 3: If X -> Y1 Y2 ... Yk,
                        // then add Follow(X) to Follow(Yk)
                        debug!(
                            "[Follow Builder] Rule 3: Push Follow({}) to Follow({})",
                            production.lhs, term
                        );
                        changed |= self.insert_follow(term, &production.lhs);

                        if !self.first_produce_epsilon(term) {
                            debug!("[Follow Builder] {} does not produce ε, break", term);
                            break;
                        }
                    } // Rule 3 checking End
                }
            }

            if !changed {
                break;
            }
        }
    }

    fn first_produce_epsilon(&self, term: &Term) -> bool {
        self.first
            .get(term)
            .map_or(false, |first| first.contains(&epsilon()))
    }

    // Insert term to Follow(x)
    ///
    /// return true if the Follow(x) changes
    /// otherwise return false
    fn insert_term(&self, x: &'grammar Term, term: &'grammar Term) -> bool {
        let mut follow = self.follow.borrow_mut();
        // Follow(x)
        let follow_x = follow.get_mut(x).unwrap();

        // Insert term to Follow(x)
        let before = follow_x.len();
        follow_x.insert(term);
        let after = follow_x.len();

        // check if set changes
        before != after
    }

    // Insert dollar to Follow(x)
    fn insert_dollar(&self, x: &'grammar Term) -> bool {
        self.insert_term(x, dollar())
    }

    // Insert set to Follow(x)
    fn insert_set(&self, x: &'grammar Term, set: HashSet<&'grammar Term>) -> bool {
        let mut follow = self.follow.borrow_mut();
        // Follow(x)
        let follow_x = follow.get_mut(x).unwrap();

        // Insert set into Follow(x)
        let before = follow_x.len();
        follow_x.extend(set);
        let after = follow_x.len();

        // check if set changes
        before != after
    }

    fn follow(&self, x: &Term) -> HashSet<&'grammar Term> {
        self.follow
            .borrow()
            .get(x)
            .map_or_else(HashSet::new, |set| set.clone())
    }

    /// Insert Follow(tx) into Follow(rx)
    pub(crate) fn insert_follow(&mut self, rx: &'grammar Term, tx: &'grammar Term) -> bool {
        // Follow(tx)
        let follow_tx = self.follow(tx);
        debug!("Insert {:?} to Follow({})", follow_tx, rx);
        // Insert Follow(tx) into Follow(rx)
        self.insert_set(rx, follow_tx)
    }

    pub(crate) fn build(mut self, start: &'grammar Term) -> Follow<'grammar> {
        self.build_follow(start);
        Follow {
            follow: self.follow.into_inner(),
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::utils::follow::builder::FollowBuilder;
    use crate::utils::follow::Follow;
    use bnf::{Grammar, Term};
    use std::collections::HashSet;

    use std::str::FromStr;

    pub fn grammar() -> Grammar {
        let input = r#"
        <E> ::= <T> <E'>
        <E'> ::= '+' <T> <E'> | 'ε'
        <T> ::= <F> <T'>
        <T'> ::= '*' <F> <T'> | 'ε'
        <F> ::= '(' <E> ')' | 'id'
        "#;
        let grammar: Grammar = input.parse().unwrap();
        grammar
    }

    fn get_follow<'a>(follow: &'a Follow, term: &str) -> HashSet<&'a Term> {
        let term = Term::from_str(term).unwrap();
        follow.follow.get(&term).unwrap().clone()
    }

    #[test]
    fn it_works() {
        let grammar = grammar();
        let start = Term::Nonterminal("E".to_string());
        let follow = FollowBuilder::new(&grammar).build(&start);

        assert_eq!(get_follow(&follow, "<E>").len(), 2);
        assert_eq!(get_follow(&follow, "<E'>").len(), 2);
        assert_eq!(get_follow(&follow, "<T>").len(), 3);
        assert_eq!(get_follow(&follow, "<T'>").len(), 3);
        assert_eq!(get_follow(&follow, "<F>").len(), 4);
    }
}
