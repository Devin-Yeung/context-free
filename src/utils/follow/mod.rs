use crate::utils::follow::builder::FollowBuilder;
use bnf::{Grammar, Term};
use itertools::Itertools;
use std::collections::{HashMap, HashSet};
use tabled::builder::Builder;
use tabled::Table;

mod builder;

pub struct Follow<'grammar> {
    pub(crate) follow: HashMap<&'grammar Term, HashSet<&'grammar Term>>,
}

impl<'grammar> Follow<'grammar> {
    pub fn new(grammar: &'grammar Grammar, start: &'grammar Term) -> Follow<'grammar> {
        FollowBuilder::new(grammar).build(start)
    }

    pub fn follow_of(&self, x: &Term) -> impl Iterator<Item = &&'grammar Term> {
        // TODO: remove the unwrap?
        self.follow.get(x).unwrap().iter()
    }

    pub fn tabled(&self) -> Table {
        let mut table = Builder::new();
        table.set_header(["Term", "Follow(X)"]);
        for (term, first) in self
            .follow
            .iter()
            .filter(|(term, _)| matches!(term, Term::Nonterminal(_)))
            .sorted_by(|a, b| a.0.cmp(b.0))
        {
            table.push_record([term.to_string(), first.iter().sorted().join(", ")]);
        }
        table.build()
    }
}

#[cfg(test)]
mod tests {
    use crate::utils::follow::Follow;
    use bnf::Term;
    use std::str::FromStr;

    #[test]
    fn it_works() {
        let grammar = r#"
        <E> ::= <T> <E'>
        <E'> ::= '+' <T> <E'> | 'ε'
        <T> ::= <F> <T'>
        <T'> ::= '*' <F> <T'> | 'ε'
        <F> ::= '(' <E> ')' | 'id'
        "#
        .parse()
        .unwrap();
        let start = Term::from_str("<E>").unwrap();
        let follow = Follow::new(&grammar, &start);
        insta::assert_display_snapshot!(follow.tabled());
    }

    #[test]
    fn test_case_1() {
        let grammar = r#"
        <P> ::= <Q> 'id' <R>
        <Q> ::= '∃' | '∀'
        <R> ::= <E> '=' <E>
        <E> ::= <T> <E'>
        <E'> ::= '+' <T> <E'> | 'ε'
        <T> ::= '(' <E> ')' | 'id'
        "#
        .parse()
        .unwrap();
        let start = Term::from_str("<P>").unwrap();
        let follow = Follow::new(&grammar, &start);
        insta::assert_display_snapshot!(follow.tabled());
    }
}
