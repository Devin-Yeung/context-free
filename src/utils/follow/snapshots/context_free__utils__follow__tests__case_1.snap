---
source: src/utils/follow/mod.rs
expression: follow.tabled()
---
+------+--------------------+
| Term | Follow(X)          |
+------+--------------------+
| <E>  | "$", ")", "="      |
+------+--------------------+
| <E'> | "$", ")", "="      |
+------+--------------------+
| <P>  | "$"                |
+------+--------------------+
| <Q>  | "id"               |
+------+--------------------+
| <R>  | "$"                |
+------+--------------------+
| <T>  | "$", ")", "+", "=" |
+------+--------------------+
